aptos-move/move-examples/marketplace/sources/coin_listing.move

```diff
@@ -0,0 +1,1080 @@
+            let Bid { bidder, coins } = option::extract(&mut auction_listing.current_bid);
+            let current_bid = coin::value(&coins);
+            coin::deposit(bidder, coins);
+            (option::some(bidder), option::some(current_bid), current_bid + auction_listing.bid_increment)
+        } else {
+            (option::none(), option::none(), auction_listing.starting_bid)
+        };
+
+        assert!(bid_amount >= minimum_bid, error::invalid_argument(EBID_TOO_LOW));
+        let coins = coin::withdraw<CoinType>(bidder, bid_amount);
+        let bid = Bid {
+            bidder: signer::address_of(bidder),
+            coins,
+        };
+        option::fill(&mut auction_listing.current_bid, bid);
+
+        let fee_schedule = listing::fee_schedule(object);
+        coin::transfer<CoinType>(
+            bidder,
+            fee_schedule::fee_address(fee_schedule),
+            fee_schedule::bidding_fee(fee_schedule, bid_amount),
+        );
+
+        let now = timestamp::now_seconds();
+        let minimum_end_time = now + auction_listing.minimum_bid_time_before_end;
+
+        if (auction_listing.auction_end_time < minimum_end_time) {
+            auction_listing.auction_end_time = minimum_end_time
+        };
+
+        let bid_event_data = BidEvent {
+            new_bidder: signer::address_of(bidder),
+            new_bid: bid_amount,
+            previous_bidder,
+            previous_bid,
+        };
+        event::emit_event(&mut auction_listing.bid_events, bid_event_data);
+    }
+
+    /// Once the current time has elapsed the auctions run time, allow the auction to be settled by
+    /// distributing out the asset to the winner or the auction seller if no one bid as well as
+    /// giving any fees to the marketplace that hosted the auction.
+    public entry fun complete_auction<CoinType>(
+        object: Object<Listing>,
+    ) acquires AuctionListing {
+        let listing_addr = listing::assert_started(&object);
+        assert!(exists<AuctionListing<CoinType>>(listing_addr), error::not_found(ENO_LISTING));
+
+        let AuctionListing {
+            starting_bid: _,
+            bid_increment: _,
+            current_bid,
+            auction_end_time,
+            minimum_bid_time_before_end: _,
+            buy_it_now_price: _,
+            bid_events,
+            purchase_event,
+        } = move_from<AuctionListing<CoinType>>(listing_addr);
+
+        let now = timestamp::now_seconds();
+        assert!(auction_end_time <= now, error::invalid_state(EAUCTION_NOT_ENDED));
+
+        let seller = listing::seller(object);
+
+        let (purchaser, price, royalties, commission)  = if (option::is_some(&current_bid)) {
+            let Bid { bidder, coins } = option::destroy_some(current_bid);
+
+            let fee_schedule = listing::fee_schedule(object);
+            let price = coin::value(&coins);
+
+            let commission_charge = fee_schedule::commission(fee_schedule, price);
```

Comment:
A lot of this logic is redundant with purchase(). Can we refactor this into a different function (e.g. purchase_internal_with_price or something)

Reply:
ahh, I remember, this result in slightly less optimal code for cases where auctions don't end with a bidder... but whatever... I think most of the outcome results in no real impact.